src/OboeDecoder.bsv

```diff
@@ -1,59 +1,28 @@
 package OboeDecoder;
+
 import OboeTypeDef::*;
 
-// Typedef: RType
-//   R-type instruction format.
-typedef struct {
-  Bit#(7) funct7;
-  ArchRegId rs2;
-  ArchRegId rs1;
-  Bit#(3) funct3;
-  ArchRegId rd;
-  Bit#(7) opcode;
-} RType deriving(Eq, Bits);
-
-// Typedef: IType
-//   I-type instruction format.
-typedef struct {
-  Bit#(12) imm;
-  ArchRegId rs1;
-  Bit#(3) funct3;
-  ArchRegId rd;
-  Bit#(7) opcode;
-} IType deriving(Eq, Bits);
-
-// Typedef: SType
-//   S-type instruction format.
-typedef struct {
-  Bit#(7) imm2;
-  ArchRegId rs2;
-  ArchRegId rs1;
-  Bit#(3) funct3;
-  Bit#(5) imm1;
-  Bit#(7) opcode;
-} SType deriving(Eq, Bits);
-
-// Typedef: UType
-//   U-type instruction format.
-typedef struct {
-  Bit#(20) imm;
-  ArchRegId rd;
-  Bit#(7) opcode;
-} UType deriving(Eq, Bits);
-
-// Typedef: OpCodeWidth
-typedef 7 OpCodeWidth;
-
-Bit#(OpCodeWidth) op_imm    = 'b0010011;
-Bit#(OpCodeWidth) op_r      = 'b0110011;
-Bit#(OpCodeWidth) op_lui    = 'b0110111;
-Bit#(OpCodeWidth) op_auipc  = 'b0010111;
-Bit#(OpCodeWidth) op_jal    = 'b1101111;
-Bit#(OpCodeWidth) op_jalr   = 'b1100111;
-Bit#(OpCodeWidth) op_branch = 'b1100011;
-Bit#(OpCodeWidth) op_load   = 'b0000011;
-Bit#(OpCodeWidth) op_store  = 'b0100011;
-Bit#(OpCodeWidth) op_csr    = 'b1110011;
+// Function: immIType
+//   Generate I-type immediate value
+function Word immIType(IType inst) = signExtend(inst.imm);
+
+// Function: immSType
+//   Generate S-type immediate value
+function Word immSType(SType inst) = signExtend({inst.imm2, inst.imm1});
+
+// Function: immUType
+//   Generate U-type immediate value
+function Word immUType(UType inst) = {inst.imm, 12'd0};
+
+// Function: immJType
+//   Generate J-type immediate value
+function Word immJType(JType inst) = signExtend({inst.imm4, inst.imm3, inst.imm2, inst.imm1, 1'b0}); 
+
+// Function: immBType
+//   Generate B-type immediate value
+function Word immBType(BType inst) = signExtend({inst.imm4, inst.imm3, inst.imm2, inst.imm1, 1'b0}); 
+
+TrapCause illegal_instruction = TrapCause {isInterrupt: False, code: 2};
```

Comment:
I think the definition of different types of trap could be placed in typedef file for better reusability.

Reply:
Sure, thank you for your valuable advice!